apps sc & wc: improve dry-run and deploy scripts

robinelastisys opened the pull request:
**What this PR does / why we need it**:
The dry-run.sh and deploy scripts we currently use will not check against the actual state of the cluster. We are now adding commando flags to enable the option of using sync if a user want to.

**Which issue this PR fixes** 
fixes #382

**Message to reviewers**
Im not sure if this is the right way to do it. This was the closest i could figure out. Feel free to point me out in the right direction if needed.

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [x] Some test is done to check if the problem still persists.

simonklb commented on bin/dry-run.bash:
```diff
@@ -21,13 +21,13 @@ if [[ $1 == "sc" ]]; then
   : "${config[config_file_sc]:?Missing service cluster config file}"
 
-  "${here}/ops.bash" helmfile sc diff
+  "${here}/ops.bash" helmfile sc template | kubectl sc diff
```
Could you give an example of what this solves?

robinelastisys replied:
Actually not sure. @pettersv Do you have a good example?

viktor-f replied:
As noted in the issue `helm diff` does not check what's actually running in the cluster, but it was suggested that piping the template to `kubectl diff` could actually provide us with this. So if there was any manual changes we could see how that would impact this upgrade.

simonklb replied:
Gotcha, I guess one should refrain from doing manual changes instead of going the route of adding things to git first etc, but I see the point and it's a good one!

Are we sure that doing `kubectl diff` instead of `helm diff` won't lose any information? It's hard to visualize, but could there be something that helm tracks that is not visible by just looking at the "plain k8s state"?

viktor-f replied:
Yes I think that everyone agrees that we should stay away from manual changes. But it will happen sometimes for different reasons.

At least I am not sure, so it would be very good to test (@robinelastisys have you been testing these changes, and if so then how have you tested it?). But I think you would get all relevant information.

simonklb commented on scripts/deploy-sc.sh:
```diff
@@ -34,6 +34,6 @@ echo "Installing helm charts" >&2
 declare -a helmfile_opt_flags
 [[ -n "$INTERACTIVE" ]] && helmfile_opt_flags+=("$INTERACTIVE")
-helmfile -f . -e service_cluster "${helmfile_opt_flags[@]}" apply --suppress-diff
+helmfile -f . -e service_cluster "${helmfile_opt_flags[@]}" sync --suppress-diff
```
Why are you doing this change?

robinelastisys replied:
This is the description of the commands:

    sync      sync all resources from state file (repos, releases and chart deps)
    apply     apply all resources from state file only when there are changes

So, i think it would be good for the sake of recovering a state? Not sure i have done changes in the right place. It was the only place i could find in where it would effect the "apply" command as the issuer wanted. Couldn't find a "helmfile apply" statement in either apps.sh or bootstrap.sh .

simonklb replied:
I'm not sure if this actually will change something in practice, but if it does it's definitely worth mentioning in the changelog as this is changing the apply command, not dry-run.

OlleLarsson replied:
The reason why this was even brought up in the first place is that helm-diff does not check the acutally state in kubernetes, and thus if somebody has made any manual changes to let's say a deployment, helm-diff won't notice that and thus not upgrade the release. sync, will always upgrade the releases to a new version thus syncing the state on every single apply.

Personally, now that we are aware of the fact of this difference between apply and sync. Whether or not we want to run sync only and never apply, I can't say.

simonklb replied:
Worth investigating before merging? :smile:

Xartos replied:
I think we said that we should use sync since this is the same as running `helm upgrade --install` (instead of `helm apply` for when apply is used) which is how you would do it if you were using helm.

simonklb replied:
`helm apply` isn't a thing I think? At least not on my end:
```
% helm apply
Error: unknown command "apply" for "helm"
Run 'helm --help' for usage.
% helm version --short
v3.2.4+g0ad800e
```

simonklb replied:
Just did a quick search since I wasn't totally aware of the commands and I think the top answer [here](https://stackoverflow.com/questions/59703760/helmfile-sync-vs-helmfile-apply) explained it pretty clearly.

To me it feels like we should not run `helmfile sync` since it's going to basically upgrade the chars even though there are no changes made. Or am I missing something here?

Xartos replied:
Oh, yea I misread the documentation there.
But I think the reason to why we want this is (as olle said) diff will not notice stuff that's manually edited. And since that might be the case that we need to do that (for an emergency fix or something) we should not allow this manual change to hang around for a couple of upgrades until the release are actually redeployed and breaks everything again (maybe months later). So in my mind it's better that we notice this change quicker when we might have a chance of remembering why. But there might be some issues with upgrading the release that I don't see?

simonklb replied:
Gotcha! I'm a bit worried that manual changes that have been done would be destroyed by a sync though? For example, what happens if you have made manual changes to some labels and it runs an upgrade of the chart release? It's all of these unknowns that I would be a bit cautious introducing before we actually know more.

If anyone here actually have made some effort into testing this and researching sync vs apply implications please add it to the changelog so that we don't introduce this without anyone knowing what is going on.

simonklb commented on WIP-CHANGELOG.md:
```diff
@@ -0,0 +1,11 @@
+
+- The dry-run and deploy scripts now checks against the state of the cluster while ran.
+
+### Fixed
+
+
+### Added
+
+
+### Removed
```
No need to add empty headings.

simonklb reviewed (changes requested):
Mostly looking for explanations. Otherwise this looks fine!

OlleLarsson commented on bin/dry-run.bash:
```diff
@@ -21,13 +21,13 @@ if [[ $1 == "sc" ]]; then
   : "${config[config_file_sc]:?Missing service cluster config file}"
 
-  "${here}/ops.bash" helmfile sc diff
+  "${here}/ops.bash" helmfile sc template | kubectl sc diff
```
```suggestion
  "${here}/ops.bash" helmfile sc template | kubectl diff
```

OlleLarsson replied:
Have you actually tried running this?
```suggestion
  "${here}/ops.bash" helmfile sc template | kubectl diff
```

OlleLarsson commented on bin/dry-run.bash:
```diff
@@ -21,13 +21,13 @@ if [[ $1 == "sc" ]]; then
   : "${config[config_file_wc]:?Missing workload cluster config file}"
 
-  "${here}/ops.bash" helmfile wc diff
+  "${here}/ops.bash" helmfile wc template | kubectl wc diff
```
Same

robinelastisys commented:
We had a meeting with the outcome:

The changes could be activated by using a command flag

robinelastisys commented:
Now i have presented a start of a solution. Not the most beautiful but it works. "apply-sync" and "dry-run-sync".

Xartos commented on bin/apps.bash:
```diff
@@ -23,23 +23,45 @@ apps_init() {
-        with_kubeconfig "${secrets[kube_config_sc]}" \
-            CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-sc.sh"
-    )
+    if [ ${#} -eq 1 ]; then
+
+        (
+            : "${scripts_path:?Missing scripts path}"
+            : "${secrets[kube_config_sc]:?Missing service cluster kubeconfig}"
+            with_kubeconfig "${secrets[kube_config_sc]}" \
+                CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-sc.sh" "${1}"
+        )
+    else
+        (
+            : "${scripts_path:?Missing scripts path}"
+            : "${secrets[kube_config_sc]:?Missing service cluster kubeconfig}"
+            with_kubeconfig "${secrets[kube_config_sc]}" \
+                CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-sc.sh"
+        )
+
+    fi
```
You should be able to simplify this to something like this, right?

```suggestion
        (
            : "${scripts_path:?Missing scripts path}"
            : "${secrets[kube_config_sc]:?Missing service cluster kubeconfig}"
            with_kubeconfig "${secrets[kube_config_sc]}" \
                CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-sc.sh" "${1:-""}"
        )
```

Xartos commented on bin/apps.bash:
```diff
@@ -23,23 +23,45 @@ apps_init() {
-        with_kubeconfig "${secrets[kube_config_wc]}" \
-            CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-wc.sh"
-    )
+    if [ ${#} -eq 1 ]; then
+
+        (
+            : "${scripts_path:?Missing scripts path}"
+            : "${secrets[kube_config_wc]:?Missing workload cluster kubeconfig}"
+            with_kubeconfig "${secrets[kube_config_wc]}" \
+                CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-wc.sh" "${1}"
+        )
+    else
+        (
+            : "${scripts_path:?Missing scripts path}"
+            : "${secrets[kube_config_wc]:?Missing workload cluster kubeconfig}"
+            with_kubeconfig "${secrets[kube_config_wc]}" \
+                CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-wc.sh"
+        )
+
+    fi
```
Same here

Xartos commented on bin/apps.bash:
```diff
@@ -75,33 +97,65 @@ apps_sc() {
-    apps_run_sc
+    #    #[ "$1" != "--skip-template-validate" ] && template_validate_sc
+
+    if [ ${#} -eq 2 ]; then
+
+        #[ "$1" != "--skip-template-validate" ] && template_validate_sc
+        apps_run_sc "$2"
+    else
+        #[ "$1" != "--skip-template-validate" ] && template_validate_sc
+        apps_run_sc
 
+    fi
```
Same here

Xartos commented on bin/apps.bash:
```diff
@@ -75,33 +97,65 @@ apps_sc() {
+    # but feels good enough until we figure out something smarter.
+    #    #[ "$1" != "--skip-template-validate" ] && template_validate_sc
+
+    if [ ${#} -eq 2 ]; then
 
+        #[ "$1" != "--skip-template-validate" ] && template_validate_wc
+        apps_run_wc "$2"
+    else
+        #[ "$1" != "--skip-template-validate" ] && template_validate_wc
+        apps_run_wc
+
+    fi
```
Same here

Xartos commented on bin/apps.bash:
```diff
@@ -75,33 +97,65 @@ apps_sc() {
     config_load "$1"
-    apps_wc "$2"
+    if [ ${#} -eq 3 ]; then
+        apps_wc "$2" "$3"
+    else
+        apps_wc "$2"
+    fi
+
 elif [[ $1 == "sc" ]]; then
     config_load "$1"
-    apps_sc "$2"
+    if [ ${#} -eq 3 ]; then
+        apps_sc "$2" "$3"
+    else
+        apps_sc "$2"
+    fi
```
And here

Xartos commented on bin/dry-run.bash:
```diff
@@ -21,14 +21,28 @@ if [[ $1 == "sc" ]]; then
   : "${config[config_file_sc]:?Missing service cluster config file}"
 
-  "${here}/ops.bash" helmfile sc diff
+  if [ ${#} -eq 2 ]; then
```
I would like that you actually verify that the variable is set to "sync" and not just assuming that it's set correctly.

This applies to all places where this assumption is done.

Xartos commented on bin/dry-run.bash:
```diff
@@ -21,14 +21,28 @@ if [[ $1 == "sc" ]]; then
+    "${here}/ops.bash" helmfile sc diff
+
+  fi
+elif
+  [[ $1 == "wc" ]]
+then
```
```suggestion
elif [[ $1 == "wc" ]]; then
```

Xartos reviewed (changes requested):
Would it be possible to instead of introducing a new command, having a flag instread? Like
```console
$ bin/ck8s apply --sync
```
I guess something like [getops](https://www.golinuxcloud.com/bash-getopts/) would be something but that doesn't support long arguments (only single letter). If it's not feasable I'm okey with this approach as well.

I would also like to rename the `dry-run-sync` to `dry-run-kubectl` since it has nothing to do with sync.
There's also some formatting issues I have commented on

robinelastisys commented:
So, now i have done a lot of changes.

1. I changed from apply-sync and dry-run-sync to --sync and --kubectl.
2. I changed the if-cases to go by "${1:-""}" instead of if-case series.
3. I did some  string checks at the entry point of apps.bash and in deploy-wc & sc for further development if/when more flags come.

robinelastisys commented on scripts/deploy-wc.sh:
```diff
@@ -5,7 +5,7 @@ set -euo pipefail
 # shellcheck disable=SC1090 # Can't follow non-constant source.
 source "${SCRIPTS_PATH}/../bin/common.bash"
-
+s
```
dammit haha :smile:

robinelastisys commented on bin/ck8s:
```diff
@@ -13,9 +13,9 @@ usage() {
     echo "  bootstrap <wc|sc>                               bootstrap the cluster" 1>&2
     echo "  apps <wc|sc> [--skip-template-validate]         deploy the applications" 1>&2
-    echo "  apply <wc|sc> [--skip-template-validate]        bootstrap and apps" 1>&2
+    echo "  apply <wc|sc> [--sync] [--skip-template-validate] bootstrap and apps" 1>&2
```
I thought about that me also.. its a tricky one as getopts aren't created to handle longer flags like that :thinking:

simonklb replied:
@robinelastisys usually long flags/options is done by iterating over the arguments, would that work here? Something like this, for example:
```
for arg in "$@"; do
  case "$arg" in
    "--sync") sync=1 ;;
    "--skip-template-validate") skip_template_validate=1 ;;
  esac
done
```

robinelastisys replied:
Hmmmmm.. interesting idea! Should work somehow indeed!  :thinking:

robinelastisys replied:
So, now i have implemented a version in where the input parameters are parsed and minor bugs are fixed.

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +71,75 @@ apps_sc() {
     #[ "$1" != "--skip-template-validate" ] && template_validate_sc
-    apps_run_sc
+
+    if [ "$#" -gt 0 ] && [ "${*: -1}" = "sync" ]; then
+        apps_run_sc "${*: -1}"
+    else
+        apps_run_sc
+    fi
```
Wouldn't this work?

```suggestion
        apps_run_sc "${*: -1}"
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +71,75 @@ apps_sc() {
     #[ "$1" != "--skip-template-validate" ] && template_validate_wc
-    apps_run_wc
+
+    if [ "$#" -gt 0 ] && [ "${*: -1}" = "sync" ]; then
+        apps_run_wc "${*: -1}"
+    else
+        apps_run_wc
+    fi
```
Same here

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +71,75 @@ apps_sc() {
     #[ "$1" != "--skip-template-validate" ] && template_validate_sc
-    apps_run_sc
+
+    if [ "$#" -gt 0 ] && [ "${*: -1}" = "sync" ]; then
```
Since we can be sure that "sync" are on position 3 I would rather have `$3` instead of this. 
Less confusion.

Xartos commented on bin/apps.bash:
```diff
@@ -23,23 +23,19 @@ apps_init() {
-        with_kubeconfig "${secrets[kube_config_sc]}" \
-            CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-sc.sh"
-    )
+    : "${scripts_path:?Missing scripts path}"
+    : "${secrets[kube_config_sc]:?Missing service cluster kubeconfig}"
+    with_kubeconfig "${secrets[kube_config_sc]}" \
+        CONFIG_PATH="${CK8S_CONFIG_PATH}" "${scripts_path}/deploy-sc.sh" "${1:-""}"
```
Why removing the parenthesis?

Xartos reviewed (changes requested):
Just some minor details

robinelastisys commented:
Done! Looks a lot smoother now indeed! :-) :+1:

Xartos commented on bin/ck8s:
```diff
@@ -43,21 +56,21 @@ case "${1}" in
-        "${here}/bootstrap.bash" "${2}"
-        "${here}/apps.bash" "${2}" "${3}"
-        ;;
+    [[ "${2}" =~ ^(wc|sc)$ ]] || usage
+    "${here}/bootstrap.bash" "${2}"
+    "${here}/apps.bash" "${2}" "$SKIP" "$SYNC"
+    ;;
```
Intentation
```suggestion
        [[ "${2}" =~ ^(wc|sc)$ ]] || usage
        "${here}/bootstrap.bash" "${2}"
        "${here}/apps.bash" "${2}" "$SKIP" "$SYNC"
        ;;
```

Xartos commented on bin/ck8s:
```diff
@@ -43,21 +56,21 @@ case "${1}" in
-        [[ "${2}" =~ ^(wc|sc)$ ]] || usage
-        "${here}/dry-run.bash" "${2}"
-        ;;
+    [[ "${2}" =~ ^(wc|sc)$ ]] || usage
+        "${here}/dry-run.bash" "${2}" $KUBECTL
+    ;;
```
Indentation
```suggestion
        [[ "${2}" =~ ^(wc|sc)$ ]] || usage
        "${here}/dry-run.bash" "${2}" $KUBECTL
        ;;
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +73,50 @@ apps_sc() {
     # at least Helmfile does "just in time" template validation. Not as nice,
     # but feels good enough until we figure out something smarter.
-    #
+    #    #[ "$1" != "--skip-template-validate" ] && template_validate_sc
+
```
This can be removed
```suggestion
    #
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +73,50 @@ apps_sc() {
 apps_wc() {
     apps_init
-    # See rationale above.
+    #
+    # The first few Charts install CRDs, which will make template validation
+    # fail. CRDs are "changes" to the Kubernetes API, hence validation against
+    # the Kubernetes API cannot be done. OTOH, manually adding the CRDs during
+    # bootstrap is error-prone and adds maintenance burden.
+    #
+    # While it would be nice to have some template validation before `helmfile apply`,
+    # at least Helmfile does "just in time" template validation. Not as nice,
+    # but feels good enough until we figure out something smarter.
+    #    #[ "$1" != "--skip-template-validate" ] && template_validate_sc
+
```
Reverted to
```suggestion
    # See rationale above.
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +73,50 @@ apps_sc() {
-
-
 if [[ $1 == "wc" ]]; then
+
```
New line not needed
```suggestion
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +73,50 @@ apps_sc() {
     config_load "$1"
-    apps_wc "$2"
+    apps_wc "$2" "$3"
+
```
New line not needed
```suggestion
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +73,50 @@ apps_sc() {
+    apps_wc "$2" "$3"
+
 elif [[ $1 == "sc" ]]; then
+
```
New line not needed
```suggestion
```

Xartos commented on bin/apps.bash:
```diff
@@ -75,34 +73,50 @@ apps_sc() {
-    apps_sc "$2"
+    apps_sc "$2" "$3"
 else
+
```
New line not needed
```suggestion
```

Xartos commented on bin/ck8s:
```diff
@@ -33,6 +33,19 @@ usage() {
     exit 1
 }
 
+
```
New line not needed
```suggestion
```

Xartos commented on bin/dry-run.bash:
```diff
@@ -21,14 +21,26 @@ if [[ $1 == "sc" ]]; then
+    "${here}/ops.bash" helmfile sc template | "${here}/ops.bash" kubectl sc diff -f -
+  else
+    "${here}/ops.bash" helmfile sc diff
+
```
New line not needed
```suggestion
```

Xartos commented on bin/dry-run.bash:
```diff
@@ -21,14 +21,26 @@ if [[ $1 == "sc" ]]; then
+    "${here}/ops.bash" helmfile wc template | "${here}/ops.bash" kubectl wc diff -f -
+  else
+    "${here}/ops.bash" helmfile wc diff
+
```
New line not needed
```suggestion
```

Xartos commented on bin/dry-run.bash:
```diff
@@ -21,14 +21,26 @@ if [[ $1 == "sc" ]]; then
+  fi
+
 else
+
```
New line not needed
```suggestion
```

Xartos commented on scripts/clean-wc.sh:
```diff
@@ -57,7 +57,7 @@ PROM_CRDS=$(
         --api-group=monitoring.coreos.com \
         -o name
-    )
+)
```
Revert to
```suggestion
    )
```

Xartos commented on scripts/deploy-sc.sh:
```diff
@@ -10,30 +10,34 @@ source "${SCRIPTS_PATH}/../bin/common.bash"
+    helmfile -f . -e service_cluster sync
+else
+    helmfile -f . -e service_cluster apply --suppress-diff
+
```
New line not needed
```suggestion
```

Xartos reviewed (changes requested):
I only have some formatting issues now but otherwise this looks great!

robinelastisys commented:
Done! :smile:

Xartos commented on bin/ck8s:
```diff
@@ -33,6 +33,18 @@ usage() {
+    "--skip-template-validate") SKIP="--skip-template-validate" ;;
+    "--sync") SYNC="sync" ;;
+    "--kubectl") KUBECTL="kubectl" ;;
+esac
```
Indentation
```suggestion
  esac
```

Xartos commented on bin/dry-run.bash:
```diff
@@ -21,13 +21,21 @@ if [[ $1 == "sc" ]]; then
+  else
+    "${here}/ops.bash" helmfile sc diff
+  fi
+elif
+  [[ $1 == "wc" ]]; then
```
This doesn't need to be split to multiple lines
```suggestion
elif [[ $1 == "wc" ]]; then
```

Xartos commented on bin/apps.bash:
```diff
@@ -76,32 +74,39 @@ apps_sc() {
-    # See rationale above.
-    #[ "$1" != "--skip-template-validate" ] && template_validate_wc
-    apps_run_wc
+    #
+    # The first few Charts install CRDs, which will make template validation
+    # fail. CRDs are "changes" to the Kubernetes API, hence validation against
+    # the Kubernetes API cannot be done. OTOH, manually adding the CRDs during
+    # bootstrap is error-prone and adds maintenance burden.
+    #
+    # While it would be nice to have some template validation before `helmfile apply`,
+    # at least Helmfile does "just in time" template validation. Not as nice,
+    # but feels good enough until we figure out something smarter.
+    #
```
This could still be reverted to
```suggestion
    # See rationale above
```

To minimize duplications

Xartos commented on bin/ck8s:
```diff
@@ -43,20 +55,20 @@ case "${1}" in
     apps)
         [[ "${2}" =~ ^(wc|sc)$ ]] || usage
-        "${here}/apps.bash" "${2}" "${3}"
+        "${here}/apps.bash" "${2}" $SKIP
```
We should try to use {} and "" for all variables
```suggestion
        "${here}/apps.bash" "${2}" "${SKIP}"
```

Xartos commented on bin/ck8s:
```diff
@@ -43,20 +55,20 @@ case "${1}" in
         [[ "${2}" =~ ^(wc|sc)$ ]] || usage
         "${here}/bootstrap.bash" "${2}"
-        "${here}/apps.bash" "${2}" "${3}"
+        "${here}/apps.bash" "${2}" "$SKIP" "$SYNC"
```
same here

Xartos commented on bin/ck8s:
```diff
@@ -43,20 +55,20 @@ case "${1}" in
     dry-run)
         [[ "${2}" =~ ^(wc|sc)$ ]] || usage
-        "${here}/dry-run.bash" "${2}"
+        "${here}/dry-run.bash" "${2}" $KUBECTL
```
same here

Xartos reviewed (changes requested):
Just the last few things :sweat_smile:

simonklb commented on WIP-CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 
 - The sc-logs-retention cronjob now runs without error even if no backups were found for automatic removal
 - Harbor Swift authentication configuration options has moved from `citycloud` to `harbor.persistence.swift`.
+- The dry-run and apply command now checks against the state of the cluster while ran.
```
Update this so that it reads as optional now since the functionality is behind optional flags.

viktor-f commented on bin/ck8s:
```diff
@@ -13,9 +13,9 @@ usage() {
     echo "  bootstrap <wc|sc>                               bootstrap the cluster" 1>&2
     echo "  apps <wc|sc> [--skip-template-validate]         deploy the applications" 1>&2
-    echo "  apply <wc|sc> [--skip-template-validate]        bootstrap and apps" 1>&2
+    echo "  apply <wc|sc> [--sync] [--skip-template-validate] bootstrap and apps" 1>&2
     echo "  test <wc|sc>                                    test the applications" 1>&2
```
```suggestion
    echo "  apply <wc|sc> [--sync] [--skip-template-validate] bootstrap and apps" 1>&2
    echo "  test <wc|sc>                                      test the applications" 1>&2
```
Nit: fix the indentation for the whole `usage` text (I can't mark all lines in this comment) so that it matches `bootstrap and apps`

viktor-f commented on bin/ck8s:
```diff
@@ -43,20 +55,20 @@ case "${1}" in
     apps)
         [[ "${2}" =~ ^(wc|sc)$ ]] || usage
-        "${here}/apps.bash" "${2}" "${3}"
+        "${here}/apps.bash" "${2}" "${SKIP}"
```
Shouldn't sync be available for this command as well?
```suggestion
        "${here}/apps.bash" "${2}" "${SKIP}" "${SYNC}"
```

viktor-f commented on bin/ck8s:
```diff
@@ -13,9 +13,9 @@ usage() {
     echo "  init                                            initialize the config path" 1>&2
     echo "  bootstrap <wc|sc>                               bootstrap the cluster" 1>&2
     echo "  apps <wc|sc> [--skip-template-validate]         deploy the applications" 1>&2
```
Shouldn't sync be available for this command as well?
```suggestion
    echo "  apps <wc|sc> [--sync] [--skip-template-validate]  deploy the applications" 1>&2
```

OlleLarsson commented:
Please resolve some of the comments that's been fixed. A bit cumbersome to overview the pr ;)

simonklb commented on bin/apps.bash:
```diff
@@ -76,32 +74,30 @@ apps_sc() {
     # at least Helmfile does "just in time" template validation. Not as nice,
     # but feels good enough until we figure out something smarter.
     #
-    #[ "$1" != "--skip-template-validate" ] && template_validate_sc
```
Keep this comment, it was commented out because we currently can't use it but not necessarily something we want to have disabled forever.

simonklb commented on bin/apps.bash:
```diff
@@ -76,32 +74,30 @@ apps_sc() {
 apps_wc() {
     apps_init
-    # See rationale above.
-    #[ "$1" != "--skip-template-validate" ] && template_validate_wc
```
Keep this comment, it was commented out because we currently can't use it but not necessarily something we want to have disabled forever.

viktor-f reviewed (approved):
LGTM, well done!

simonklb reviewed (approved):
LGTM! Nice :)

Xartos reviewed (approved):
LGTM! Great work
